### DISCUSSION TITLE
Update the Dictionary and related classes to work similar to ValueTuple with regards to element names.

### DISCUSSION THREAD

**Discussion by vsfeedback** [dotnet/csharplang#6718] [Language Ideas] [answered]
_This issue has been moved from [a ticket on Developer Community](https://developercommunity.visualstudio.com/t/Update-the-Dictionary-and-related-classe/1631869)._

---
I absolutely LOVE the ValueTuple class in .NET.   One of the absolute best things is that i can provide names for the elements of the tuple.
For example, i can do this:

```
Dim myTuple as (Name as String, HireDate as DateTime)
Dim tupleName as String = myTuple.Name
```
The valuable feature of the above is:  myTuple.**Name**
The '. Name' part of the tuple versus just 'Item1' (which still works).

However, when working with Dictionary, or any other similar class with a Key and Value we don't have that.

It would be AMAZING to be able to do the following:

```
Dim betterDict as New System.Collections.Generic.Dictionary(Of Name as String, HireDate as DateTime)
Dim nameOfFirstElement as String = betterDict.First(). Name
```
The request is to be able to do the above part:  betterDict.First().**Name**

When analyzing code with dictionary objects, you're stuck with only 'Key' and 'Value' and if there's no comments to describe what those elements are, you have to chase it down in code somewhere else.

---
#### Feedback Bot on 1/9/2022, 11:29 PM: 

(private comment, text removed)

**Comment by terrajobst** [comment]
I don't think there is much we can do in the class libraries to address this as a `Dictionary<TKey, TValue>` isn't using any tuples. There is some minor integration between types like `KeyValuePair<TKey, TValue>` which can be deconstructed like `var (x, y) = kvp;` (not sure what the VB syntax is).

What you're asking is really a language feature that would let you create an alias for the generic type parameter and all the places it is exposed, which is basically how this works on value tuples. The challenge is that this isn't a single place; it would conceptually have to be any place that provides a property or field of the type of the generic parameter and it would have to be tracked across types, for example, when you enumerate a dictionary we return an `IEnumerable<KeyValuePair<TKey, TValue>>`. I'm not sure how feasible this is to be honest.

**Reply by terrajobst** [reply]
@KathleenDollard shouldn't this be moved to vblang instead?

**Comment by KathleenDollard** [comment, marked as answer]
We are not active on VBLang, so do not move it there. I think we need to effectively close this (the issue in Feedback is already closed). Also, if we did this, it should be a C# feature as well. But...this would be a massive undertaking that is out of scope unless we were to rethink generics. 

Let's effectively close this as a cool idea that we do not know how we could ever implement.

**Comment by TahirAhmadov** [comment]
Please see:
https://github.com/dotnet/csharplang/discussions/269
https://github.com/dotnet/csharplang/discussions/894
https://github.com/dotnet/csharplang/discussions/1883
https://github.com/dotnet/csharplang/discussions/5706
https://github.com/dotnet/csharplang/discussions/5835
